### PR TITLE
Fix Add Series modal performance issues by moving the logic to the BE

### LINKED ISSUE #28045: Add series modal pulls all cards and viz settings and it's unusable on big instances
**Describe the bug**
Saw this on a customer with lots of cards: the add series modal takes at least a minute to return results, and when it does, it's so massive that the browser has problems rendering the list.

If you try using it now on Stats, you'll see a 40 seconds wait and the response is 2.8Mb (Feb-2023), which is massive. We should paginate the results at least

**Logs**
NA

**To Reproduce**
1) Go to stats and create a dashboard, then add a question and click on the "add series" modal

**Expected behavior**
We should bring smaller chunks and render those, as there's a search

**Screenshots**
NA

**Information about your Metabase Installation:**
- Your browser and the version: Brave

**Severity**
P2

**Additional context**
Reported by a customer

### PR DESCRIPTION
**Pull request by qnkhuat**
Closes https://github.com/metabase/metabase/issues/28045

## Description

Fixes the add series modal that allows combining cards on dashboards. The logic that determines if two cards are compatible was implemented on the FE, it needed both cards and their metadata. So previously the add series modal loaded all the cards from the instance which made it unusable on larger instances. It initially hid structured incompatible questions, but for native queries, users had to try to select them to understand if they are compatible which was not great.

In this PR we added a new BE endpoint that exposes only compatible cards with a cursor-based pagination to avoid performance issues.

## How to verify

- Create a few compatible structured/native cards that have numeric/date dimensions for the x-axis
- Ensure you can combine them on a dashboard
- Ensure the results are paginated (to simplify testing this you can change the page size in `AddSeriesModal/QuestionList.tsx`

## API

Adding a new endpoint `GET /api/card/:id/series` with arguments:
- `last_cursor` : id of the last card from the last page
- `limit`: page_size
- `query`: filter cards by name
- `exclude_ids`: a list of card ids to exclude 
The endpoint will return only compatible cards to be used as a series for the given card.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30321)

### CODE REVIEW

File: test/metabase/api/card_test.clj [hunk 1]
```diff
@@ -388,6 +389,212 @@
...
+        (is (false? (api.card/series-are-compatible? scalar-1 scalar-2-cols)))
+        (is (false? (api.card/series-are-compatible? scalar-2-cols scalar-2))))
+
+      (testing "can only be cominbed with scalar cards"
+        (is (false? (api.card/series-are-compatible? scalar-1 line-card)))))))
+
+
+
+
+
```
Comment (qnkhuat) [inline test/metabase/api/card_test.clj:597]:
```suggestion
```

File: src/metabase/api/card.clj [hunk 2]
```diff
@@ -207,6 +209,153 @@
...
+                                 second-card
+                                 (get-in second-card [:visualization_settings :graph.metrics]))]
+         (cond
+           ;; must have at least one dimension and one metric
+           (or (zero? (count new-dimensions))
+             (zero? (count new-metrics)))
+           false
+
+           ;; all metrics must be numeric
+           (not (every? lib.types.isa/numeric? new-metrics))
```
Comment (qnkhuat) [inline src/metabase/api/card.clj:241]: .

**Comment by alxnddr** [comment]
@qnkhuat I merged my branch into yours to make developing and testing easier. FE should be working at this moment.
On the BE side seems like we need to add the support of:
- `search` param for searching questions by name
- `exclude` param to exclude already selected card ids

**Comment by qnkhuat** [comment]
@alxnddr for search you can use `query` to search by `card.name`.

I just added `exclude_ids` to exclude a list of card by ids.

**Review by alxnddr — COMMENTED** [review]
@qnkhuat I pushed some FE specs and we already had e2e which fails currently because of a regression in the compatibility check logic. Can you please have a look and also port these specs from the FE to BE?
https://github.com/metabase/metabase/pull/30321/files#diff-64904e39c6b8d26d7972546f3f67c4aefdbdd911eef2fae336a2aa909e40d967

Besides that, it feels great compared to the previous implementation that loaded everything 👏

**Comment by qnkhuat** [comment]
nothing was wrong with our logic 🎉 , the test question was created without viz-settings and we need it to infer compatible series correctly. I fixed the test data and should works now dc347ce2b3fdad192392ef30f1be99078c2f53df

File: test/metabase/http_client.clj [hunk 3]
```diff
@@ -238,8 +243,13 @@
     (cond-> parsed
       ;; un-nest {:request-options {:request-options <my-options>}} => {:request-options <my-options>}
       (:request-options parsed) (update :request-options :request-options)
-      ;; convert query parameters into a flat map [{:k :a, :v 1} {:k :b, :v 2}] => {:a 1, :b 2}
-      (:query-parameters parsed) (update :query-parameters (partial into {} (map (juxt :k :v)))))))
+      ;; convert query parameters into a flat map [{:k :a, :v 1} {:k :b, :v 2} {:k :b, :v 3}] => {:a 1, :b [2 3]}
+      (:query-parameters parsed) (update :query-parameters (fn [query-params]
```
Comment (qnkhuat) [inline test/metabase/http_client.clj:247]: this allows we to do

```
(mt/http-user-http ... :param1 "a" :param1 "b")
```

which is equivalent of `?param1=a&param1=b` in the URL.

**Review by npfitz — APPROVED** [review]
Functionality works well, and the new code is very easy to follow 🥳. Bit of a different pattern than we usually do with the list component hitting the API directly, but the call is light enough that it shouldn't be a problem, and the search / pagination feels great.

File: src/metabase/api/card.clj [hunk 4]
```diff
@@ -207,6 +209,175 @@
       (when-not (Boolean/parseBoolean ignore_view)
         (events/publish-event! :card-read (assoc <> :actor_id api/*current-user-id*))))))
 
+(defn- card-columns-from-names
+  [card names]
+  (when-let [names (set names)]
+    (filter #(names (:name %)) (:result_metadata card))))
+
+(defmulti series-are-compatible?
```
Comment (noahmoss) [inline src/metabase/api/card.clj:217]: Can this be moved down closer to the method implementations?

File: src/metabase/api/card.clj [hunk 5]
```diff
@@ -207,6 +209,175 @@
...
+  "Check if the `second-card` is compatible to be used as series of `card`."
+  (fn [card _second-card]
+   (:display card)))
+
+(defn- cols->kebab-case
+  [cols]
+  (map #(update-keys % u/->kebab-case-en) cols))
+
+(defn- area-bar-line-serie-are-compatible?
```
Comment (noahmoss) [inline src/metabase/api/card.clj:226]:
```suggestion
(defn- area-bar-line-series-are-compatible?
```

File: src/metabase/api/card.clj [hunk 6]
```diff
@@ -207,6 +209,175 @@
...
+                 (lib.types.isa/date? (first new-dimensions)))
+           false
+
+           ;; both or neither primary dimension must be numeric
+           ;; a timestamp field is both date and number so don't enforce the condition if both fields are dates; see #2811
+           (and (not= (lib.types.isa/numeric? (first initial-dimensions))
+                      (lib.types.isa/numeric? (first new-dimensions)))
+                (not (and
+                       (lib.types.isa/date? (first initial-dimensions))
+                       (lib.types.isa/date? (first new-dimensions)))))
```
Comment (noahmoss) [inline src/metabase/api/card.clj:262]: nit
```suggestion
                (not (and
                      (lib.types.isa/date? (first initial-dimensions))
                      (lib.types.isa/date? (first new-dimensions)))))
```

**Review by noahmoss — APPROVED** [review]
not familiar with the original series-related logic so I trust that's been ported accurately, but I went over the code and just have some nits

File: src/metabase/api/common.clj [hunk 7]
```diff
@@ -610,9 +610,32 @@
          (reconcile-position-for-collection! old-collection-id old-position nil)
          (reconcile-position-for-collection! new-collection-id nil new-position))))))
 
+
+;;; ------------------------------------------ PARAM PARSING FNS ----------------------------------------
+
 (defn bit->boolean
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/api/common.clj:616]: It's being used in api.collection and api.search so I think it's okay to leave it there.

File: src/metabase/api/card.clj (hunk 6 above)
Comment (qnkhuat) [reply src/metabase/api/card.clj]: a9ef5737f0

File: src/metabase/api/card.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/api/card.clj]: a9ef5737f0

File: src/metabase/api/card.clj (hunk 4 above)
Comment (qnkhuat) [reply src/metabase/api/card.clj]: yep a9ef5737f0

**Comment by paoliniluis** [comment]
guys this is FAST

**Review by dragonsahead — APPROVED** [review]

**Pull request merged**